src/js/common/schemaform/widgets/ArrayCountWidget.jsx

```diff
@@ -0,0 +1,67 @@
+import React from 'react';
+
+// We're expanding or contracting the array based on the count
+// and returning undefined if the array should be empty
+function getValue(count, value = []) {
+  if (!count || parseInt(count, 10) < 1) {
+    return undefined;
+  }
+
+  const intCount = parseInt(count, 10);
+  if (intCount < value.length) {
+    return value.slice(0, intCount);
+  }
+
+  return value.concat(Array(intCount - value.length).fill({}));
+}
+
+export default class ArrayCountWidget extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { value: !!props.value ? props.value.length : undefined };
+  }
+
+  // Need to keep the count state in sync, but don't want to overwrite if the
+  // user has chosen '' or 0
+  componentWillReceiveProps(props) {
+    const arrayData = props.value || [];
+    if (arrayData.length && this.state.value !== arrayData.length) {
+      this.setState({ value: arrayData.length });
+    } else if (!this.state.value && !!arrayData.length) {
+      this.setState({ value: arrayData.length });
+    }
+  }
+
+  updateArrayLength = (event) => {
+    const count = event.target.value;
+    this.setState({ value: count }, () => {
```

Comment:
I'm not sure if you mean the `setState` call or the callback.

For the `setState` call: I won't necessarily know what the value in the input element is if I do that in a lifecycle method. The reason for using local state here is that I can track what a user has entered into the text box separately from the length of the array field being changed. If we don't do that, the text field acts weird and won't let you delete a 0.

For the callback: I think this makes the chain of events clearer. If I put the `onChange` call in a lifecycle method, I have to be careful about only calling it when I know it's necessary, which I think is harder code to read (sort of like the `componentWillReceiveProps` method in here, which isn't immediately clear, but I don't have a choice).